src/Tools/dotnet-monitor/Egress/AzureBlobEgressFactory.cs

```diff
@@ -0,0 +1,117 @@
+using Microsoft.Diagnostics.Monitoring.RestServer;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tools.Monitor
+{
+    internal class AzureBlobEgressFactory : EgressFactory
+    {
+        private readonly ILoggerFactory _loggerFactory;
+
+        public AzureBlobEgressFactory(ILoggerFactory loggerFactory)
+            : base(loggerFactory.CreateLogger<AzureBlobEgressFactory>())
+        {
+            _loggerFactory = loggerFactory;
+        }
+
+        public override bool TryCreate(
+            string providerName,
+            IConfigurationSection providerSection,
+            Dictionary<string, string> egressProperties,
+            out ConfiguredEgressProvider provider)
+        {
+            var options = providerSection.Get<ConfigurationOptions>();
+
+            if (string.IsNullOrEmpty(options.AccountKey) &&
+                !string.IsNullOrEmpty(options.AccountKeyName))
+            {
+                if (egressProperties.TryGetValue(options.AccountKeyName, out string key))
+                {
+                    options.AccountKey = key;
+                }
+            }
+
+            if (string.IsNullOrEmpty(options.SharedAccessSignature) &&
+                !string.IsNullOrEmpty(options.SharedAccessSignatureName))
+            {
+                if (egressProperties.TryGetValue(options.SharedAccessSignatureName, out string signature))
+                {
+                    options.SharedAccessSignature = signature;
+                }
+            }
+
+            if (!TryValidateOptions(options, providerName))
+            {
+                provider = null;
+                return false;
+            }
+
+            provider = new Provider(options, _loggerFactory);
+            return true;
+        }
+
+        private class Provider : ConfiguredEgressProvider
+        {
+            private readonly ILoggerFactory _loggerFactory;
+            private readonly AzureBlobEgressProviderOptions _options;
+
+            public Provider(AzureBlobEgressProviderOptions options, ILoggerFactory loggerFactory)
+            {
+                _options = options;
+                _loggerFactory = loggerFactory;
+            }
+
+            public override async Task<EgressResult> EgressAsync(
+                Func<CancellationToken, Task<Stream>> action,
+                string fileName,
+                string contentType,
+                IEndpointInfo source,
+                CancellationToken token)
+            {
+                // TODO: Add metadata based on source
+                var streamOptions = new AzureBlobEgressStreamOptions();
+                streamOptions.ContentType = contentType;
+
+                var provider = new AzureBlobEgressProvider(_options, _loggerFactory.CreateLogger<AzureBlobEgressProvider>());
```

Comment:
This can be cached as a readonly field and be reused for multiple egress invocations because the provider options will not change for each invocation. The ConfiguredEgressProvider implementation will be recreated by EgressConfigureOptions if the underlyding provider options change.

Reply:
Done